File: 3782.last-remaining-integer-after-alternating-deletion-operations.py
#
# @lc app=leetcode id=3782 lang=python3
#
# [3782] Last Remaining Integer After Alternating Deletion Operations
#

# @lc code=start
class Solution:
    def lastInteger(self, n: int) -> int:
        head = 1
        step = 1
        remaining = n
        left = True
        
        while remaining > 1:
            # When going from right and count is even, first element is deleted
            if not left and remaining % 2 == 0:
                head += step
            
            step *= 2
            remaining = (remaining + 1) // 2
            left = not left
        
        return head
# @lc code=end

File: 3782.last-remaining-integer-after-alternating-deletion-operations.cpp
#
# @lc app=leetcode id=3782 lang=cpp
#
# [3782] Last Remaining Integer After Alternating Deletion Operations
#
# @lc code=start
class Solution {
public:
    long long lastInteger(long long n) {
        return solve(n, true);
    }
    
private:
    long long solve(long long n, bool leftToRight) {
        if (n == 1) return 1;
        
        if (leftToRight) {
            return 2 * solve(n / 2, false) - 1;
        } else {
            return 2 * solve(n / 2, true) - (n & 1);
        }
    }
};
# @lc code=end